### PR TITLE
Inventory service 1

### PR DESCRIPTION
**Pull request by pjazdzewski1990**
ASAP

### CODE REVIEW

File: _posts/2015-06-16-inventory-service-1.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+* Consistent - after each transaction the data is left in an consistent state, it doesn't matter whether the transaction ended with a success or was aborted
+
+* Isolated - two different operations won't interfere with each other
+
+* Durable - changes done by successful transactions are visible after commit
+
+As you can see the key thing here is `transaction`. A transaction bundles operations together and is applied in an orderly fashion to the rows in the db. STM brings those concepts to the code. No more threads racing to access shared memory. With STM all modifications have to be done via transactions so the ACID properties hold for our concurrent code. You still might think "why should I care? I don't have any shared mutable state in my Scala programs". In some sense most of us have a mutable state ... we call it a database. 
+
+So where does Activate fit here? Activate is build on top of RadonSTM (which brings STM to Scala) and aims to simplify the persistence layer by applying the STM approach to data handling. It might seem really confusing at the beginning, but trust me you will quickly get it after looking on some code samples. 
+
```
Comment (marioosh) [inline _posts/2015-06-16-inventory-service-1.md]: `RadonSTM` as link to project please :)

File: _posts/2015-06-16-inventory-service-1.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+      .createTable(
+        _.column[String]("name"),
+        _.column[String]("code"),
+        _.column[OfficeEntity]("inOffice"))
+  }
+}
+{% endhighlight %}
+[\[SchemaMigration.scala\]](https://github.com/pjazdzewski1990/Inventory-App/blob/master/src/main/scala/io/scalac/inventory/db/activate/SchemaMigration.scala)
+
+As you can see nothing out of the ordinary here. We define an "setup" function with Activate DSL and add a timestamp defining execution order for this migration.
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: I believe it is supposed to be `We define a "setup" function...` - just a typo.

File: _posts/2015-06-16-inventory-service-1.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+I didn't use them in my examples, but there are also for comprehensions. Slick uses them extensively. In Activate you can use them, too. Just take a look on this [extension.](https://github.com/fwbrasil/activate/blob/v1.5/activate-docs/query.md#slick-queries) In fact you don't have to choose Activate or Slick, you can use them in combination.
+
+### Session handling
+
+Both Activate and Slick provide us with means for managed and automatic session handling. Ad hoc sessions are handled by `withDynSession` in Slick and in Activate the default `transactional` block creates one when needed. In both libraries session are objects, so it can be created and passed around when a managed solution is needed. 
+
+### Speed (of writing)
+
+Slick comes with a price when you have to pay at the beginning of the development, when defining your base classes. Activate skips this part almost completely, hence it feels faster to write.
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: I would use `Slick comes with a price that you have to pay...`.

**Comment by marioosh** [comment]
:+1:

File: _posts/2015-06-16-inventory-service-1.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+## <a name="stm" href="#stm">What is STM?</a>
+
+Wikipedia [defines it](https://en.wikipedia.org/wiki/Software_transactional_memory) as follows: software transactional memory (STM) is a concurrency control mechanism analogous to database transactions for controlling access to shared memory in concurrent computing. It is an alternative to lock-based synchronization. Hmmm... but what does it mean in practice? 
+
+From a programmers perspective STM is a way of handling concurrent reads and writes to a shared memory. This technique was inspired by relational databases. As we all know most of these dbs have ACID guarantees, which means that their operations are 
+
+* Atomic - transactions are done or aborted as a whole, they can't be half-way done
+
+* Consistent - after each transaction the data is left in an consistent state, it doesn't matter whether the transaction ended with a success or was aborted
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: `in an consistent` -> `in a consistent`

File: _posts/2015-06-16-inventory-service-1.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+As you can see the key thing here is `transaction`. A transaction bundles operations together and is applied in an orderly fashion to the rows in the db. STM brings those concepts to the code. No more threads racing to access shared memory. With STM all modifications have to be done via transactions so the ACID properties hold for our concurrent code. You still might think "why should I care? I don't have any shared mutable state in my Scala programs". In some sense most of us have a mutable state ... we call it a database. 
+
+So where does Activate fit here? Activate is build on top of [RadonSTM](https://github.com/fwbrasil/radon-stm) (which brings STM to Scala) and aims to simplify the persistence layer by applying the STM approach to data handling. It might seem really confusing at the beginning, but trust me you will quickly get it after looking on some code samples. 
+
+## <a name="repository" href="#repository">Activate components explained</a>
+
+Now I will guide you through the process of creating a simple repository class using Activate. In case you get lost take a look at the source code [here](https://github.com/pjazdzewski1990/Inventory-App)
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: Missing colon at the end of the sentence.

File: _posts/2015-06-16-inventory-service-1.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+### Migration
+
+Now let's prepare the db to use it with the app. In order to use that we will need to run migration to setup all the tables and fields. This part is also very straightforward 
+
+<script src="https://gist.github.com/pjazdzewski1990/9fac7c1ac344b2fd266b.js"></script>
+[\[SchemaMigration.scala\]](https://github.com/pjazdzewski1990/Inventory-App/blob/master/src/main/scala/io/scalac/inventory/db/activate/SchemaMigration.scala)
+
+As you can see nothing out of the ordinary here. We define a "setup" function with Activate DSL and add a timestamp defining execution order for this migration.
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: should probably be `Activate's DSL`?

File: _posts/2015-06-16-inventory-service-1.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+### Migration
+
+Now let's prepare the db to use it with the app. In order to use that we will need to run migration to setup all the tables and fields. This part is also very straightforward 
+
+<script src="https://gist.github.com/pjazdzewski1990/9fac7c1ac344b2fd266b.js"></script>
+[\[SchemaMigration.scala\]](https://github.com/pjazdzewski1990/Inventory-App/blob/master/src/main/scala/io/scalac/inventory/db/activate/SchemaMigration.scala)
+
+As you can see nothing out of the ordinary here. We define a "setup" function with Activate DSL and add a timestamp defining execution order for this migration. 
+
+Uff that was a bit of code
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: Miissing colon.

File: _posts/2015-06-16-inventory-service-1.md [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+Uff that was a bit of code 
+
+## <a name="working" href="#working">Working with Activate</a>
+
+Having all the code in place we can start doing some real work. Let's define `ActivateRepository`. The code goes roughly as follows:
+
+<script src="https://gist.github.com/pjazdzewski1990/9d0373424c0cd30f94f1.js"></script>
+[\[ActivateRepository.scala\]](https://github.com/pjazdzewski1990/Inventory-App/blob/master/src/main/scala/io/scalac/inventory/db/activate/ActivateRepository.scala)
+
+Let's dig a bit in the code. The simplest to explain is probably `createItem`. It leverages the power of Activate. In fact if you look closely it simply creates a new `ItemEntity` object. The magic happens behind the scenes. Activate propagates the data we put into the instance to the db. We work with plain objects, Activate does all the heavy lifting (as long as we remember about the `transactional` block). Simple as that. This nice model frees you of thinking about the backing db details and help you concentrate on code.
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: I believe it should be `...and helps you...`.

File: _posts/2015-06-16-inventory-service-1.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+Having that in mind understanding the rest is easy. Although `updateItem` looks intimidating it uses the same trick, but this time we need to load the original record, load the new data and update the item with it. You might wonder about `map`, `canonical` and `flatten`, but they are just noise to keep the types in check.
+
+`readItem` and `deleteItem` are a bit different. To make use of them we will use Activate's type safe query mechanism. Here's an example
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: Missing colon.

File: _posts/2015-06-16-inventory-service-1.md [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+<script src="https://gist.github.com/pjazdzewski1990/9c3c35ee036eed5b9956.js"></script>
+
+As you can see the return and the query itself are typed. We make usage out of it when comparing (Activate defines operators starting with :, like :== for comparison). Framework defines a lot more features than that. You can explore them all [in the docs](https://github.com/fwbrasil/activate/blob/v1.5/activate-docs/query.md)
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: Missing dot.

File: _posts/2015-06-16-inventory-service-1.md [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+## <a name="slick" href="#slick">Comparing Activate with Slick</a>
+
+We can't simply compare Activate and Slick. Although they both "live" in the persistence layer, they don't aim to do the same thing. In my opinion Slick tries to be only an functional mapper from relational databases to code, where Activate a fully featured framework for persistence. Despite those differences I would like to share with you some thoughts that came to my mind when implementing the same repository with Slick.
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: `an functional` -> `a functional`

File: _posts/2015-06-16-inventory-service-1.md [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+### Migrations
+
+Slick don't help us much in this case. Although have access to `TableQuery.ddl` we need to come up with a scheme for using this on our own. Slick doesn't help us in this regard. Activate on other hand provides us with a standard way of doing schema versioning using `Migration` base class. It's all done in code. It reminds me of [evolutions](https://www.playframework.com/documentation/2.0/Evolutions) in Play framework.
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: `Slick don't` -> `Slick doesn't`

File: _posts/2015-06-16-inventory-service-1.md [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+### Table definitions
+
+In Slick when modelling a table you have to define a `Table[T]` class. It has to define all the columns and some meta-data. This `Table` class is the backbone of type safe query model. Although handy, it might be a bit confusing at the beginning. You have to remember to fill the Tuple with all the column types, define the `*` projection and add meta-data (like `O.AutoInc`). A lot for a beginner ... but after you get a grip on Slick it gets natural. For Activate `Entity` is the table. Every single class extending `Entity` corresponds to a db table and every instance of that class to a row. No special syntax is required. In case you need to add an meta information to the table, you can add it as a function in the class. Neat
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: I would use `It has to contain all the columns...` because the word `define` is being used extensively here. But if it has to be `define` just let it as is.

File: _posts/2015-06-16-inventory-service-1.md [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+
+### Querying
+
+When it comes to querying Slick gives us the `TableQuery` class. The part of Slick queries I really like is the fact, that they are backed by objects, so you can pass them around and compose easily. Activate on the other hand relies on functions. You pass a function to `query` and the database gets queried. I feel like in this case function provide us with less flexibility than objects.
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: I feel the need for a comma: `When it comes to querying, Slick...`

File: _posts/2015-06-16-inventory-service-1.md [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+
+I didn't use them in my examples, but there are also for comprehensions. Slick uses them extensively. In Activate you can use them, too. Just take a look on this [extension.](https://github.com/fwbrasil/activate/blob/v1.5/activate-docs/query.md#slick-queries) In fact you don't have to choose Activate or Slick, you can use them in combination.
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: As it is a recent pass, more appropriate would be `Although, I haven't used them in my examples, there are [specify what there is] for comprehensions.`

File: _posts/2015-06-16-inventory-service-1.md [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+
+### Session handling
+
+Both Activate and Slick provide us with means for managed and automatic session handling. Ad hoc sessions are handled by `withDynSession` in Slick and in Activate the default `transactional` block creates one when needed. In both libraries session are objects, so it can be created and passed around when a managed solution is needed.
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: `session are objects`? was it supposed to be `sessions are objects`? or maybe reword a bit

File: _posts/2015-06-16-inventory-service-1.md [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+
+Both Activate and Slick provide us with means for managed and automatic session handling. Ad hoc sessions are handled by `withDynSession` in Slick and in Activate the default `transactional` block creates one when needed. In both libraries session are objects, so it can be created and passed around when a managed solution is needed. 
+
+### Speed (of writing)
+
+Slick comes with a price that you have to pay at the beginning of the development, when defining your base classes. Activate skips this part almost completely, hence it feels faster to write.
+
+### Speed (of execution)
+
+At this point I can't tell you much about it. Can't wait to do some benchmarks ... When I do that I will make sure to put the results on our blog.
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: Maybe it's fine to omit the word `that` from `When I do that I will make sure...`

File: _posts/2015-06-16-inventory-service-1.md [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+
+The full Slick example is available [on Github.](https://github.com/pjazdzewski1990/Inventory-App/blob/master/src/main/scala/io/scalac/inventory/db/slick/SlickRepository.scala)
+
+## <a name="summary" href="#summary">Lessons learned and opinions</a>
+
+To sum up. To me Activate looks really promising and I can't wait to test it in a bigger problem. It really liked the simplicity of Activate and how it removed the boilerplate from my code.
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: `It really liked` -> `I really liked`?

File: _posts/2015-06-16-inventory-service-1.md [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+
+Notable mentions - interesting things that Activate offers, but didn't make to this post:
```
Comment (mareksuscak) [inline _posts/2015-06-16-inventory-service-1.md]: i would use `interesting things that didn't make it to this post but Activate offers`

File: _posts/2015-06-16-inventory-service-1.md (hunk 6 above)
Comment (pjazdzewski1990) [inline _posts/2015-06-16-inventory-service-1.md]: In fact, it's not a DSL at all. I will just drop it
